src/backend/env/post.rs

```diff
@@ -622,20 +641,31 @@ impl Post {
     }
 
     pub fn count(state: &State) -> usize {
-        state.posts.len() + state.memory.posts.len()
+        state.posts.len()
+            + POSTS.with(|p| match p.borrow().as_ref() {
+                Some(tree) => tree.len(), _ => 0
+            }) as usize
+            // TODO: delete this line after migration
+            + state.memory.posts.len()
     }
 
     // Get the post from the heap if available, or load from the stable memory into the cache and
     // return the reference to it
     pub fn get<'a>(state: &'a State, post_id: &PostId) -> Option<&'a Post> {
         state.posts.get(post_id).or_else(|| {
             let boxed = cache().get(post_id).or_else(|| {
-                state.memory.posts.get(post_id).and_then(|mut post: Post| {
-                    let cache = cache();
-                    post.archived = true;
-                    cache.insert(*post_id, Box::new(post));
-                    cache.get(post_id)
-                })
+                // TODO: delete state.memory access after migration
+                state
+                    .memory
+                    .posts
+                    .get(post_id)
```

Comment:
How do I check that this `get` will access the heap instead of the stable memory?

Reply:
I was thinking to do a round of tests with println's which would disclose the source of data. I even wanted to put this into PR but was afraid that it will blow up the change with debugging helpful for local testing _only_. I definitely can add it if you find it useful.

Reply:
I might be missing something. Is the idea to migrate posts from stable memory to stable structures? Or do we load all cold posts into heap and then migrate?

Reply:
I don't see any special handling to make sure that the old stable memory doesn't overlap with the new stable memory for stable structure. So then my first question stands. This must not access stable memory because that would be broken, right?

Reply:
In `post_upgrade` we move everything into heap _and_ initialize structs. That's it.

The actual cold-archiving into the **new structs** happens then on am hourly timer (or using `archive` if needed).

So this particular `get` function will never be called before `post_upgrade` has finished. But after it, `state.memory.posts` struct's indices will all be empty as the cold-archiving will not touch it anymore and posts were moved out. So it cannot ever try to read the stable memory using the old allocator. Instead, we'll try the next line, which will already read from the new structs.

Reply:
@ulan yes, you are correct. What you seem to miss is that by moving posts from stable memory to the heap, the current allocator stops tracking the corresponding `post_id`. So if we moved all posts, the current allocator will not have any memory coordinates to resolve to to read anything. Does this make sense now?

Reply:
Ok, if the function is either unreachable or a no-op, why do we have it here?

Reply:
I just tried to keep the change as small as possible to make it easier to review and reason about but it seems like I achieved the opposite 😂

Reply:
In general I agree about minimizing changes to simplify review. This particular case relies on a subtle property. If that goes wrong for some reason, we will get corrupted data. I would remove this to make this case robust.

Reply:
The reason is that as soon as we initialize stable structs, all accesses to the old allocator are potentially dangerous.

Reply:
Absolutely agree, I was not really aware of this subtle context in my head. Will remove the line!

Reply:
Removed.